File: frontend/src/types/autonomy.ts
export type AutonomyMode = 'plan' | 'suggest' | 'auto'

export interface AutonomyProfile {
  id: number
  name: string
  mode: AutonomyMode
  description: string | null
  is_default: boolean
  allowed_tools: string[] | null
  denied_tools: string[] | null
  max_file_size_kb: number | null
  require_approval_for: string[] | null
  created_at: string
  updated_at: string
}

export interface ActiveAutonomy {
  mode: AutonomyMode
  profile_name: string
  description: string | null
}

export const AUTONOMY_MODES: Record<AutonomyMode, { label: string; description: string; color: string }> = {
  plan: {
    label: 'Plan',
    description: 'Read-only — agent inspects but cannot modify files',
    color: 'bg-blue-100 text-blue-800 dark:bg-blue-900 dark:text-blue-300',
  },
  suggest: {
    label: 'Suggest',
    description: 'Interactive — agent proposes changes, you approve each',
    color: 'bg-amber-100 text-amber-800 dark:bg-amber-900 dark:text-amber-300',
  },
  auto: {
    label: 'Auto',
    description: 'Full autonomy — agent auto-approves all actions',
    color: 'bg-emerald-100 text-emerald-800 dark:bg-emerald-900 dark:text-emerald-300',
  },
}

File: backend/app/models/autonomy.py
"""ORM model for agent autonomy settings."""
from datetime import datetime, timezone

from sqlalchemy import String, Boolean, DateTime, JSON
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base


def _utcnow() -> datetime:
    return datetime.now(timezone.utc)


class AutonomyProfile(Base):
    """Per-session or global autonomy profile.

    Modes:
      - plan:     read-only, agent can inspect but not modify
      - suggest:  agent proposes changes, user approves each
      - auto:     full autonomy, agent auto-approves all actions
    """

    __tablename__ = "autonomy_profiles"

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String(128), unique=True)
    mode: Mapped[str] = mapped_column(String(16), default="suggest")  # plan | suggest | auto
    description: Mapped[str | None] = mapped_column(String(512), nullable=True)
    is_default: Mapped[bool] = mapped_column(Boolean, default=False)
    allowed_tools: Mapped[list | None] = mapped_column(JSON, nullable=True)
    denied_tools: Mapped[list | None] = mapped_column(JSON, nullable=True)
    max_file_size_kb: Mapped[int | None] = mapped_column(nullable=True)
    require_approval_for: Mapped[list | None] = mapped_column(JSON, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=_utcnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=_utcnow, onupdate=_utcnow)
